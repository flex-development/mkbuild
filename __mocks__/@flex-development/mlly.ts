/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
