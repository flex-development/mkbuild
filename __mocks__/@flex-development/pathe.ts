/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
