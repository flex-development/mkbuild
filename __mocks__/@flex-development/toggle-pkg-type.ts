/**
 * @file Mocks - @flex-development/toggle-pkg-type
 * @module mocks/flex-development/toggle-pkg-type
 * @see https://github.com/flex-development/toggle-pkg-type
 */

export default vi.fn()
