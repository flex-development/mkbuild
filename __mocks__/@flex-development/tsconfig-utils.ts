/**
 * @file Mocks - @flex-development/tsconfig-utils
 * @module mocks/flex-development/tsconfig-utils
 * @see https://github.com/flex-development/tsconfig-utils
 */

/**
 * [`@flex-development/tsconfig-utils`][1] module type.
 *
 * [1]: https://github.com/flex-development/tsconfig-utils
 */
type Actual = typeof import('@flex-development/tsconfig-utils')

/**
 * `@flex-development/tsconfig-utils` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>(
  '@flex-development/tsconfig-utils'
)

export const ScriptTarget = actual.ScriptTarget
export const normalizeCompilerOptions = vi.fn(actual.normalizeCompilerOptions)
export const normalizeModuleResolution = vi.fn(actual.normalizeModuleResolution)
export const resolvePaths = vi.fn(actual.resolvePaths)
