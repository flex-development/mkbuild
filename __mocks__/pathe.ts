/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
