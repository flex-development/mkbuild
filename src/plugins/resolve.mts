/**
 * @file Plugins - resolve
 * @module mkbuild/plugins/resolve
 */

import VIRTUAL_MODULE_RE from '#internal/virtual-module-re'
import type WithUndefined from '#types/with-undefined'
import { isBuiltin } from '@flex-development/is-builtin'
import type {
  ResolveIdOptions,
  ResolveOptions
} from '@flex-development/mkbuild'
import {
  isRelativeSpecifier,
  resolveAlias,
  resolveModule,
  toRelativeSpecifier,
  type ResolveModuleOptions
} from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedInputOptions,
  PartialResolvedId,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:resolve'

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Plugin API.
 *
 * @see {@linkcode ResolveModuleOptions}
 *
 * @internal
 *
 * @extends {ResolveModuleOptions}
 */
interface Api extends ResolveModuleOptions {
  /**
   * URL of root directory.
   *
   * @override
   */
  cwd: URL
}

/**
 * Create a module resolver plugin.
 *
 * @todo `makeAbsoluteExternalsRelative` support (?)
 *
 * @internal
 *
 * @see {@linkcode Api}
 * @see {@linkcode Plugin}
 * @see {@linkcode ResolveOptions}
 *
 * @this {void}
 *
 * @param {ResolveOptions | null | undefined} [options]
 *  Module resolve options
 * @return {Plugin<Api>}
 *  Module resolver plugin
 */
function plugin(
  this: void,
  options?: ResolveOptions | null | undefined
): Plugin<Api> {
  /**
   * Plugin api.
   *
   * @const {Api} api
   */
  const api: Api = { ...options } as Api

  return {
    api,
    buildStart,
    name: PLUGIN_NAME,
    resolveId
  }

  /**
   * Initialize the plugin api.
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {undefined}
   */
  function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): undefined {
    ok(typeof options.root === 'string', 'expected `root`')

    api.cwd = pathe.pathToFileURL(options.root)

    api.extensions ??= [
      '.mjs',
      '.js',
      '.mts',
      '.ts',
      '.jsx',
      '.tsx',
      '.json'
    ]

    api.fs = options.fs

    return void this
  }

  /**
   * Resolve `id`.
   *
   * > ðŸ‘‰ **Note**: Does nothing if `id` follows virtual module naming
   * > conventions.
   *
   * @see https://rollupjs.org/plugin-development/#resolveid
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {string} id
   *  The module specifier to resolve
   * @param {string | undefined} importer
   *  The parent module path
   * @param {ResolveIdOptions} options
   *  Hook options
   * @return {Promise<PartialResolvedId | undefined>}
   *  Resolve result
   * @throws {unknown}
   */
  async function resolveId(
    this: PluginContext,
    id: string,
    importer: string | undefined,
    options: ResolveIdOptions
  ): Promise<PartialResolvedId | undefined> {
    /**
     * Resolve result.
     *
     * @var {WithUndefined<PartialResolvedId> | undefined} resolved
     */
    let resolved: WithUndefined<PartialResolvedId> | undefined

    if (!VIRTUAL_MODULE_RE.test(id)) {
      ok(api.cwd, 'expected `api.cwd`')

      /**
       * URL of parent module.
       *
       * @const {URL} parent
       */
      const parent: URL = importer ? pathe.pathToFileURL(importer) : api.cwd

      if (isBuiltin(id)) {
        resolved = {
          external: true,
          id: String(await resolveModule(id, parent))
        }
      } else {
        /**
         * Specifier of aliased module.
         *
         * @const {string | null} url
         */
        const aliased: string | null = resolveAlias(id, {
          ...api,
          absolute: false,
          parent
        })

        // reset `id` if path alias match was found
        id = aliased ?? id

        /**
         * URL of resolved module.
         *
         * @const {URL} url
         */
        const url: URL = await resolveModule(id, parent, api)

        resolved = {
          external: !options.isEntry,
          id: isRelativeSpecifier(id) ? toRelativeSpecifier(url, parent) : id
        }
      }
    }

    return resolved as PartialResolvedId
  }
}
