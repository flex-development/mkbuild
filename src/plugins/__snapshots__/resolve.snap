// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:plugins/resolve > should return plugin 1`] = `
{
  "api": {},
  "buildStart": [Function buildStart],
  "name": "mkbuild:resolve",
  "resolveId": [Function resolveId],
}
`;
